arrow2_convert/src/deserialize.rs

```diff
@@ -48,23 +48,36 @@ pub trait ArrowDeserialize: ArrowField + Sized {
     /// something like for<'a> &'a T::ArrayType: IntoIterator<Item=Option<E>>,
     /// However, the E parameter seems to confuse the borrow checker if it's a reference.
     fn arrow_deserialize_internal(
```

Comment:
@aldanor Cleaning this up will need a bit more thought. Elements in arrow2 arrays are optional, since they can be nullable. Therefore when converting from Option<ArrowType> -> T (where T is non-nullable), we should check that the arrow array doesn't have a validity bitmap set in this case (otherwise the conversion should just fail upfront). If we can make the deserialize of required fields take `ArrowType` instead of `Option<ArrowType>`, that may be much cleaner.

Reply:
Option/internal stuff aside, being able to take `ArrowType` instead of `Option<ArrowType>` would be a great move and make API a lot cleaner in general, without having to mess with options and manual `.map()` all over the place (also making it more accessible for people who just want to write their own serializers/deserializers).

Reply:
Thanks for the feedback @aldanor. I started this discussion in arrow2 (https://github.com/jorgecarleitao/arrow2/discussions/1273). Lets see where that lands. I don't think this should hold up wrapping up GATs and your generics change. We can make another pass after those changes before v0.4.0 release.